File: distinct-subsequences/hua_dp.java
/*Time complexity: O(|s| * |t|)
Space complexity: O(|s| * |t|)
huahua dp cpp, runtime error
shaun translate to java, accepted
Runtime: 4 ms, faster than 95.21% of Java online submissions for Distinct Subsequences.
Memory Usage: 34.5 MB, less than 85.51% of Java online submissions for Distinct Subsequences.
*/
class Solution {
    public int numDistinct(String s, String t) {
        int ls = s.length(), lt = t.length();
        int [][] dp = new int[lt+1][ls+1];
        //fill(begin(dp[0]), end(dp[0]), 1);  
        for (int j = 0; j <= ls; j++)
            dp[0][j] = 1;
        
        for (int i  = 1; i <= lt; i++) {
            for (int j = 1; j <= ls; j++) {
                dp[i][j] = dp[i][j - 1] + (t.charAt(i - 1) == s.charAt(j - 1) ? dp[i - 1][j - 1] : 0);
            }
        }
        return dp[lt][ls];
        
    }
}


/*

class Solution {
public:
  int numDistinct(string s, string t) {
    int ls = s.length();
    int lt = t.length();
    vector<vector<int>> dp(lt + 1, vector<int>(ls + 1));
    fill(begin(dp[0]), end(dp[0]), 1);        
    for (int i = 1; i <= lt; ++i)
      for (int j = 1; j <= ls; ++j)
        dp[i][j] = dp[i][j - 1] + (t[i - 1] == s[j - 1] ? dp[i - 1][j - 1] : 0);        
    return dp[lt][ls];
  }
};
*/

File: interleaving-string/hua_dp.cpp
/*
hua dp
Runtime: 8 ms, faster than 87.25% of C++ online submissions for Interleaving String.
Memory Usage: 9 MB, less than 100.00% of C++ online submissions for Interleaving String.

*/
class Solution {
public:
  bool isInterleave(string s1, string s2, string s3) {
    int l1 = s1.length();
    int l2 = s2.length();
    int l3 = s3.length();
    if (l1 + l2 != l3) return false;
	//key point: s1.len + s2.len == s3.len
    // dp[i][j]: whehter s3[0:i+j] is a interleva of s1[0:i] and s2[0:j]
    vector<vector<int>> dp(l1 + 1, vector<int>(l2 + 1));
    dp[0][0] = true;
    for (int i = 0; i <= l1; ++i)
      for (int j = 0; j <= l2; ++j) {        
        if (i > 0) dp[i][j] |= dp[i - 1][j] && s1[i - 1] == s3[i + j - 1]; //take a char from s1, or take 1 char from s2
        if (j > 0) dp[i][j] |= dp[i][j - 1] && s2[j - 1] == s3[i + j - 1];        
      }
    return dp[l1][l2];
  }
};

